src/runtime_src/core/common/drv/kds_core.c

```diff
@@ -371,29 +373,36 @@ kds_del_cu_context(struct kds_sched *kds, struct kds_client *client,
 		return -EINVAL;
 	}
 
-	/* TODO: finish ERT abort process later */
-	if (!kds->ert_disable)
+	/* Before close, make sure no remain commands in CU's queue. */
+	outstanding = client->s_cnt[cu_idx] - client->c_cnt[cu_idx];
+	if (!outstanding)
 		goto skip;
 
-	/* Before close, make sure no remain commands in CU's queue.
-	 * There is no reason to sleep 500ms :)
-	 */
-	while (xrt_cu_abort(cu_mgmt->xcus[cu_idx], client) == -EAGAIN)
-		msleep(500);
+	kds_warn(client, "%ld outstanding command(s) on CU(%d)",
+		 outstanding, cu_idx);
 
+	if (!kds->ert_disable)
+		kds->ert->abort(kds->ert, client, cu_idx);
+	else
+		xrt_cu_abort(cu_mgmt->xcus[cu_idx], client);
+
+	/* sub-device that handle command should do abort with a timeout */
 	do {
-		msleep(100);
-		state = xrt_cu_abort_done(cu_mgmt->xcus[cu_idx]);
-	} while (!state);
+		msleep(500);
+		outstanding = client->s_cnt[cu_idx] - client->c_cnt[cu_idx];
+	} while(outstanding);
 
-	if (state == CU_STATE_BAD) {
-		kds_info(client, "CU(%d) hangs, please reset device", cu_idx);
-		/* No lock to protect bad_state.
-		 * If a new client doesn't get the updated status and send commands,
-		 * those commands would failed with TIMEOUT state.
-		 */
+	if (!kds->ert_disable) {
+		kds->ert->abort_done(kds->ert, client, cu_idx);
+		bad_state = kds->ert->is_bad_state(kds->ert);
+	} else {
+		xrt_cu_abort_done(cu_mgmt->xcus[cu_idx], client);
+		bad_state = xrt_cu_is_bad_state(cu_mgmt->xcus[cu_idx]);
+	}
+
+	if (bad_state) {
 		kds->bad_state = 1;
```

Comment:
where do we start aborting all cmds on this device once the state becomes bad?

Reply:
In xocl_kds.c, xocl_command_ioctl(). If state becomes bad, the execbuf ioctl would directly return -EDEADLK without allocating xcmd.